### PR TITLE
[Narwhal] speed up batch processing of fetched certificates

### PR DESCRIPTION
**Pull request by mwtian**
## Description 

Currently, certificates fetched during catchup are processed and accepted one-by-one, same as normal certificate handling. This turns out to be a throughput bottleneck for catching up. This PR includes one major and a few minor changes to improve the catchup throughput.

Main change: take the lock on `state` once per batch of certificates in `process_certificate_with_lock()`, instead of once per certificate. `tokio::sync::Mutex` seems to be really slow as the lock on `state`, taking 5 ~ 10s for 2000 lock operations.

Other changes:
- Increase channel size in Narwhal across the board from 1k to 10k, to avoid filling the channels too often. I did not see very noticeable memory usage increase.
- Avoid using blocking thread for verifying user signatures in transactions, which should be relatively fast.
- Cleanups.

### Private testnet

In catchup experiments with 5000 TPS and 150 validators, this seems to improve catchup speed from ~2/round to ~8/round.

Before:
Catching up after 1 hr of downtime never finished within the epoch:

image

image

After:
Catching up after 1 hr of downtime took ~20 min:

image

image

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: narwhal/primary/src/synchronizer.rs [hunk 1]
```diff
@@ -156,19 +181,6 @@ impl Inner {
 
         let digest = certificate.digest();
 
-        // Validate that certificates are accepted in causal order.
-        // This should be relatively cheap because of certificate store caching.
-        if certificate.round() > self.gc_round.load(Ordering::Acquire) + 1 {
```
Comment (akichidis) [inline narwhal/primary/src/synchronizer.rs:161]: That check saved the day a couple of times during refactoring which made simtests fail - do we see high performance penalty leaving it?

**Review by akichidis — APPROVED** [review]
Overall looks good to me, great speed up @mwtian ! A few comments before merging 👍

File: narwhal/primary/src/synchronizer.rs (hunk 1 above)
Comment (mwtian) [reply narwhal/primary/src/synchronizer.rs:161]: After the tokio Mutex fix, I think we can keep this!

File: narwhal/primary/src/synchronizer.rs [hunk 2]
```diff
@@ -811,78 +982,91 @@ impl Synchronizer {
...
-        if inner.certificate_store.contains(&digest)? {
-            debug!("Skip processing certificate {:?}", certificate);
+        let certificates = certificates
+            .into_iter()
+            .filter(|c| {
+                if inner.certificate_store.contains(&c.digest()).unwrap() {
+                    debug!("Skip processing certificate {:?}", c);
+                    inner.metrics.duplicate_certificates_processed.inc();
+                    return false;
+                }
```
[the remark this answers is not in the document]
Comment (mwtian) [reply narwhal/primary/src/synchronizer.rs:1005]: Good point!

File: narwhal/primary/src/synchronizer.rs [hunk 3]
```diff
@@ -811,78 +988,91 @@ impl Synchronizer {
...
+                // Re-check if the certificate has been suspended, which can happen before the lock is
+                // acquired.
+                if let Some(notify) = state.check_suspended(&digest) {
+                    trace!("Certificate {digest:?} is still suspended. Skip processing.");
+                    inner
+                        .metrics
+                        .certificates_suspended
+                        .with_label_values(&["dedup_locked"])
+                        .inc();
+                    return Err(DagError::Suspended(notify));
```
[the remark this answers is not in the document]
Comment (mwtian) [reply narwhal/primary/src/synchronizer.rs:1036]: Added comment and logic to try processing all certificates.

**Pull request merged**